### DISCUSSION TITLE
Render OCR as PDF

### DISCUSSION THREAD

**Discussion by ambirdev** [emgucv/emgucv#737] [General]
I see the PDFRenderer class as part of the OCR package but there are no examples on how to use it.  I want to be able to create a searchable PDF from a performed OCR.  Can anyone point at sample code to create a PDF from OCR?

**Comment by ambirdev** [comment]
The output to PDF still appears to be broken with the latest release.  Looking at the test case code:

  bool success = true;
                using (PDFRenderer pdfRenderer = new PDFRenderer("abc", Tesseract.DefaultTesseractDirectory, false))
                {
                    success &= pdfRenderer.BeginDocument("testing");
                    success &= pdfRenderer.AddImage(ocr);
                    success &= pdfRenderer.EndDocument();
                    //EmguAssert.IsTrue(success, "failed to export pdf");
                }
                if (success)
                {
                    FileInfo fi1 = new FileInfo("abc.pdf");

                }

BeginDocument fails as well as the rest of the calls to the renderer.  We would like to proceed with a purchase of the professional license but we need this functionality to work for us.  Please advise.  We tested the code on Windows/.NET 7.0(and 6.0 as well).

**Reply by emgucv** [reply]
The unit test in the 4.6.0 release here:
https://github.com/emgucv/emgucv/blob/4.6.0/Emgu.CV.Test/AutoTestOcr.cs#L38

produce this searchable PDF file:

[abc.pdf](https://github.com/emgucv/emgucv/files/10406301/abc.pdf)

It is working as expected.

**Reply by ambirdev** [reply]
We know what the issue is for us:  it is successful as a WinForms app/.NET 7.0.   It fails when the app is .NET Maui.  So the variable is .NET Maui; it is not functional.   We understand the .NET Maui functionality is experimental.  What is odd is that all other features of Emgu/OCR are functional except for this one piece under Maui.

**Reply by emgucv** [reply]
You mention that it failed on MAUI, what is the target? Is it net6.0-windows10.0.19041.0? For windows, the library should behave the same in MAUI / WinForms, the native binary are the same.

**Reply by ambirdev** [reply]
Tried with both .NET 6 and .NET 7 as a Maui app. Both fail.  It creates the PDFRenderer but this line returns a false, as well as the next two lines:

success &= pdfRenderer.BeginDocument("testing");

We dropped your test code into a bare bones .NET app(nothing more than one dialog with a button in it to launch the test code).   PDF creation works when the app is WinForms/.NET 7.0, but fails under .NET Maui on .NET 6 or 7.  Same OCR/PDF code for both Maui and Winform apps(literally cut and pasted your sample code/class into our project), just project and UI code is different.  All compiled using VS2022 with the latest version of .NET and VS.  Very odd.  We also built another Maui test app where the OCR/PDF code is in a class library that is .NET standard.  That fails as well being called from a .NET Maui app(just the PDF code, all other emgu image processing and OCR work perfectly).  Happy to share our code if it helps.

**Reply by emgucv** [reply]
Please shared the code. 

One of the important item to check is that, the 2nd parameter passed to the constructor of "PDFRenderer" must be a folder that is accessable by the app, and contains the PDF font file "pdf.ttf". If not, the PDFRender will fail.

**Comment by ambirdev** [comment]
Attached is the sample code.  You will see a Winforms app that does work and a Maui app that fails.

**Reply by emgucv** [reply]
Couldn't find the sample code, can you upload it from the GitHub issue tracker? Replying the email with attachment may not work.

**Comment by ambirdev** [comment]
[](url)
[emgu_examples.zip](https://github.com/emgucv/emgucv/files/10570374/emgu_examples.zip)

**Reply by emgucv** [reply]
Project has compilation error. Can you check if it is missing files?

![image](https://user-images.githubusercontent.com/2035816/221383925-d273337a-fd9f-42b4-b098-17d3aa65a696.png)

Thanks.

**Comment by ambirdev** [comment]
[MauiApp4.zip](https://github.com/emgucv/emgucv/files/11023663/MauiApp4.zip)

New project file to show error.  Updated to Maui 7.0.  Same failure of PDF.   Note:  run as Windows build rather than compile.  The compile throws the static main error because it's trying to create a Mac Catalyst build and I've torn that out of the project.  Would seem to be a bug by MS.  But I've tested what is in the zip file and it will run as a Windows build.  Run the app and click on the Click Me button to execute the emgu/OCR code.

**Reply by ambirdev** [reply]
Any update on this issue?

**Reply by emgucv** [reply]
Your code try to write an "abc.pdf" file into the same folder of the executable in MAUI.  

For MAUI apps, the folder of the executable is not writable. You need to write the file to some where else. (FYI, for unit test, writing pdf file into the folder of executable is fine.)

Changing this code:

```
                        using (PDFRenderer pdfRenderer = new PDFRenderer("abc", Tesseract.DefaultTesseractDirectory, false))
                        {
                            success &= pdfRenderer.BeginDocument("testing");
                            success &= pdfRenderer.AddImage(ocr);
                            success &= pdfRenderer.EndDocument();
                            //EmguAssert.IsTrue(success, "failed to export pdf");
                        }
                        if (success)
                        {

                            FileInfo fi1 = new FileInfo("abc.pdf");

                        }
```
into this
```
                        using (PDFRenderer pdfRenderer = new PDFRenderer(  Path.Join(Tesseract.DefaultTesseractDirectory , "abc"), Tesseract.DefaultTesseractDirectory, false))
                        {
                            success &= pdfRenderer.BeginDocument("testing");
                            success &= pdfRenderer.AddImage(ocr);
                            success &= pdfRenderer.EndDocument();
                            //EmguAssert.IsTrue(success, "failed to export pdf");
                        }
                        if (success)
                        {

                            FileInfo fi1 = new FileInfo(Path.Join(Tesseract.DefaultTesseractDirectory, "abc.pdf"));
                        }
```
And the abc.pdf file is correctly produced in the Tesseract.DefaultTesseractDirectory folder.

**Reply by ambirdev** [reply]
Perfect, this works on Maui.

**Comment by ambirdev** [comment]
Closed

**Comment by ambirdev** [comment]
There is one issue we have discovered:  PdfRenderer.AddImage, if the image passed in to the Tesseract engine is RGB color(24 or 32 bit) the call to AddImage returns a false.   The PDF is still generated with the detected text but the image is missing from the PDF.  Is this a bug, I assume this is not limited to adding 1 channel gray images only.  In the sample code if you generate the Mat img object with 3 channels instead of 1 you will get a failure of Addimage.

This is the line that returns a false:
success &= pdfRenderer.AddImage(ocr);
